src/Features/Core/Portable/ChangeSignature/ChangeSignatureCodeActionOperation.cs

```diff
@@ -0,0 +1,55 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Threading;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.Notification;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+
+namespace Microsoft.CodeAnalysis.ChangeSignature
+{
+    /// <summary>
+    /// Defines the <see cref="CodeActionOperation"/> for the <see cref="ChangeSignatureCodeAction"/>
+    /// This is used instead of <see cref="ApplyChangesOperation"/> as we need to show a confirmation
+    /// dialog to the user before applying the change.
+    /// </summary>
+    internal class ChangeSignatureCodeActionOperation : CodeActionOperation
+    {
+        public Solution ChangedSolution { get; }
+
+        public string? ConfirmationMessage { get; }
+
+        public ChangeSignatureCodeActionOperation(Solution changedSolution, string? confirmationMessage)
+        {
+            ChangedSolution = changedSolution ?? throw new ArgumentNullException(nameof(changedSolution));
+            ConfirmationMessage = confirmationMessage;
+        }
+
+        internal override bool ApplyDuringTests => true;
+
+        public override void Apply(Workspace workspace, CancellationToken cancellationToken)
+            => this.TryApply(workspace, new ProgressTracker(), cancellationToken);
+
+        /// <summary>
+        /// Show the confirmation message, if available, before attempting to apply the changes.
+        /// </summary>
+        internal override bool TryApply(
+            Workspace workspace, IProgressTracker progressTracker, CancellationToken cancellationToken)
+        {
+            if (ConfirmationMessage != null)
+            {
+                var notificationService = workspace.Services.GetRequiredService<INotificationService>();
+                if (!notificationService.ConfirmMessageBox(ConfirmationMessage, severity: NotificationSeverity.Warning))
```

Comment:
this is a bit weird.  how does thi work during preview?  won't this pop up and cause all sorts of issues?

Reply:
@CyrusNajmabadi interesting answer-

So from testing, in normal VS without these changes I cannot get preview to show up at all for change signature invoked via a code action.  I think that's probably a bug, unless it was disabled for code actions for some other reason.  If it should be showing preview, I can make a separate preview handling method here to not show the notification, but I'll have to dig tomorrow to figure out why preview isn't working in roslyn master first.

If I invoke via the command handling (cntrl R, cntrl V) then what happens is the same as before this change - the confirmation box pops up before the preview box (and hitting no cancels)

Reply:
> I think that's probably a bug

https://github.com/dotnet/roslyn/issues/4608

Reply:
I investigated this a bit today.  So it looks like the code action was never written with handling preview in mind.

I attempted to implement preview via the [IPreviewDialogService](http://sourceroslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/Host/IPreviewDialogService.cs,9668af67b8157caa), but it lives in the editor features layer while the code refactoring provider (+ service and operations) live in Features.  So that definition needs to move down or I need something to move the code action impl up to editor features (via some service as I don't want to entirely move change signature service up).  

I think that is a bit tangential to this PR so I'd like to move forward with the current behavior here (dialog shows when applying the code action as before).